src/main/java/servlets/FetchOptionsServlet.java

```diff
@@ -0,0 +1,58 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+ 
+package com.google.sps.servlets;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Set;
+import com.google.sps.data.TestRecipe;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+ 
+/** Servlet responsible for fetching all possible search options,
+    based on existing recipes in Datastore (i.e. recipes on Shef). */
+@WebServlet("/fetch-options")
+public class FetchOptionsServlet extends HttpServlet {
+ 
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Recipe").addSort("timestamp", SortDirection.DESCENDING);
```

Comment:
Optional: Did you have any thoughts on how to make this less expensive if we had many recipes?

I was thinking about some form of caching mechanism or storing the string set separately. I think simply setting caching options in the httpservletresponse may help per-user caching at a minimum, but I haven't explored this much.

Reply:
Please refer to Design Doc (Detailed Design, Alternatives) for my wonderful exploration into caching. TL;DR, settled on using HttpServletResponse cache-control after reading documentation for a variety of methods, got confused by how exactly cache-control works, asked a VSE, did the debugs, and then figured out that the lack of caching even with the correct placement + cache-control specification may be due to browser specifics. I could try fixing this by fiddling with eTags, but there are some requirements (detailed in the Alternatives section) that may be too time-consuming, so for the time being, I'm moving on with just the response.setHeader fix.